### PR TITLE
Enable Blazor Server and WebAssembly components to run in the same document

### LINKED ISSUE #48032: Ability to run Blazor Server and WebAssembly components side-by-side in the same document
As part of the "Full stack web UI with Blazor" effort, one of our goals is having the ability to interleave Blazor WebAssembly and Server components in the same page, e.g., have a Server component render a WebAssembly component or vice versa. One critical aspect of this feature is allowing Blazor Server and WebAssembly components to coexist on the same page in the first place.

Part of this work will include:
* Updating Blazor's JS API to take into account the possibility of multiple .NET runtimes being available. For example, we would need to resolve ambiguities around where JS->.NET invocations run when WebAssembly and Server runtimes are both available.
* Allowing Blazor Server/WebAssembly functionality to be initialized and disposed on demand. For example, if all Blazor Server components get removed from the page, we would have the option to shut down the circuit and spin it back up later if a Server component gets rendered again.

A deliberate non-goal is allowing completely separate Blazor apps to exist on the same page and behave in isolation. That piece of work is tracked separately in #38128 and addresses different use cases.

### PR DESCRIPTION
**Pull request by MackinnonBuck**
# Enable Blazor Server and WebAssembly components to run in the same document

This PR enables common interactivity scenarios for SSR'd Blazor apps with components rendered in "Server" and/or "WebAssembly" render modes.

There is still some remaining work and open questions to be resolved before Blazor Server and WebAssembly can run on the same page in complete harmony. Most of these can be addressed in future PRs:
- [ ] Update browser-side navigation to take into account the possibility of multiple .NET runtimes being present.
- [ ] Update JS initializer behavior to account for multiple runtimes (e.g., should there be multiple calls to `afterStarted()`?).
- [ ] Handle spinning up .NET runtimes after the initial page load.
- [ ] Handle closing circuits when there are no server components remaining.
- [ ] Implement "auto" mode logic.
- [ ] Figure out a good testing strategy - how can we be sure each feature works correctly when using Server and WebAssembly at the same time?
    * Probably easiest to pick out the scenarios that we know are tricky and test those.

Fixes #48032

### CODE REVIEW

File: src/Components/Components.slnf
Comment (MackinnonBuck) [inline src/Components/Components.slnf:1]: Feel free to ignore the changes to the `BlazorUnitedApp` sample. This was just to make it easy to test changes. I may later revert the commit with changes to the sample project.

File: src/Components/Web.JS/src/Boot.Server.Common.ts
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Boot.Server.Common.ts:23]: Before this PR, we had `DotNet.jsCallDispatcher` to dispatch into JS, and `DotNet.invokeMethod{Async}` to call static .NET methods. I've introduced a new `ICallDispatcher` to replace both of these.

Each `ICallDispatcher` is associated with a `DotNetCallDispatcher`, which each platform provides an implementation of (this part hasn't changed). When a .NET->JS call occurs, we then know which .NET runtime to "return" to, based on the `ICallDispatcher` that initiated the JS function call. `DotNetObject` has been extended to include a reference to its `ICallDispatcher` so it knows which .NET runtime to invoke into.

File: src/Components/Web.JS/src/Boot.Web.ts
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Boot.Web.ts:1]: This file is still very much a WIP.

File: src/Components/Web.JS/src/GlobalExports.ts
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/GlobalExports.ts:72]: `receiveDotNetDataStream` is unique in that it needs to know which `ICallDispatcher` initiated the call. So, I've added these two methods, and each of them passes in its respective `ICallDispatcher` instance to `receiveDotNetDataStream`. I'm trying to avoid having any platform-agnostic APIs that can't handle multiple platforms being available at the same time.

File: src/Components/Web.JS/src/Rendering/WebRendererInteropMethods.ts
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Rendering/WebRendererInteropMethods.ts:18]: Need to think about whether this is the right approach. Could we get away with keeping the existing behavior where `afterStarted()` only runs once (after the first renderer gets attached)?

File: src/Components/Web.JS/src/Rendering/WebRendererInteropMethods.ts
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Rendering/WebRendererInteropMethods.ts:23]: Before this PR, this method accepted a `rendererId` that was allocated on the .NET side. I've flipped this so that the `rendererId` is allocated in the browser and returned to .NET. That way, you don't have different .NET runtimes both starting at `0` and overwriting each other.

File: src/Components/Web.JS/src/Services/ComponentDescriptorDiscovery.ts
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Services/ComponentDescriptorDiscovery.ts:153]: I noticed that, for every valid component comment, there is also a component comment with the same `prerenderId` but no other information, e.g.: `<!--Blazor:{"prerenderId":"825916660abd46248dd9d28f894e203b"}-->`. This causes this method to fail. @SteveSandersonMS is this behavior expected?

File: src/Components/Web.JS/src/Services/NavigationManager.ts
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Services/NavigationManager.ts:1]: So far, the changes to this file are just fixing linting issues. I'll probably be adding more later to support navigation scenarios.

File: src/Components/Web/src/WebRenderer.cs
Comment (MackinnonBuck) [inline src/Components/Web/src/WebRenderer.cs:80]: My other comment explains the context behind this change. I've made it so that the `RendererId` is allocated on the client side.

Fun fact: Before this PR, the mechanism assigning `RendererId`s was broken. The JS interop call passing `RendererId` to the browser was initiated in the constructor, which runs before the derived constructors run. So, no matter what value was assigned to `RendererId` in derived classes, the `RendererId` passed to JS was always 0.

File: src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts
Comment (MackinnonBuck) [inline src/JSInterop/Microsoft.JSInterop.JS/src/src/Microsoft.JSInterop.ts:125]: We could make it so that calling this method throws an error unless explicitly allowed via some sort of option. That would encourage people to use `DotNetObjectReference` instead.

File: src/Components/Web.JS/src/Services/ComponentDescriptorDiscovery.ts
Comment (MackinnonBuck) [reply src/Components/Web.JS/src/Services/ComponentDescriptorDiscovery.ts:153]: Resolving because this was discussed offline.

File: src/Components/Web.JS/src/Boot.Web.ts
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Boot.Web.ts:34]: Currently, "auto" mode will cause an exception to be thrown because we don't currently handle the case where a component has multiple descriptors outputted.

File: src/Components/Web.JS/src/Services/NavigationManager.ts
Comment (MackinnonBuck) [reply src/Components/Web.JS/src/Services/NavigationManager.ts:1]: Update: this will come in a later PR.

File: src/Components/Samples/BlazorUnitedApp/Program.cs
Comment (javiercn) [inline src/Components/Samples/BlazorUnitedApp/Program.cs:32]: Could we avoid modifying the sample? This is adding work that we'll have to undo as we plug in additional stuff.

File: src/Components/Web/src/WebRenderer.cs
Comment (javiercn) [inline src/Components/Web/src/WebRenderer.cs:43]: If we want to avoid this dance, we can change the Renderer ID for a GUID instead of an int.

Comment (javiercn) [reply src/Components/Web/src/WebRenderer.cs:43]: I would avoid doing this inside the constructor and instead have a static factory method that we can call to initialize the renderer. If we need to / want to, we can have an interface for it, but I don't think we need it. It avoids the whole task "dance".

Additionally, for Blazor server we can pass in the ID directly on the call to Blazor Start. 

We could similarly do something with web assembly, maybe with an environment variable or similar that we could read from config (Plugging in the environment variables provider). I believe this is possible since we have "harmonized" the boot process.

File: src/Components/Web.JS/src/Boot.Server.Common.ts
Comment (javiercn) [inline src/Components/Web.JS/src/Boot.Server.Common.ts:76]: Could we make this a bit more idiomatic?

In this multi dispatching world I would have expected for us to new up a dispatcher instance and configure it. With a server implementation and a webassembly implementation.

File: src/Components/Web.JS/src/Boot.Web.ts
Comment (javiercn) [inline src/Components/Web.JS/src/Boot.Web.ts:43]: I would have hoped that we don't do this this way, and that instead we do an import("...") and call start on the result.

That has important implications for bundling, as we don't want to load the JS/Server code if its not used.

File: src/Components/Web.JS/src/Rendering/WebRendererInteropMethods.ts
Comment (javiercn) [reply src/Components/Web.JS/src/Rendering/WebRendererInteropMethods.ts:18]: The initializer can already detect the platform by introspecting what it is being passed to it. We need to think of what pattern apps will follow when using initializers in this world, and whether we want to support doing things like having "platform" specific initializers via ".server.js" and ".browser.js".

**Review by javiercn — COMMENTED** [review]
Looks like a great start, I have left some feedback!

File: src/Components/Samples/BlazorUnitedApp.WebAssembly/Program.cs
Comment (SteveSandersonMS) [reply src/Components/Samples/BlazorUnitedApp.WebAssembly/Program.cs:12]: Out of interest, how come it's necessary to do this anyway? I thought we didn't trim application assemblies.

File: src/Components/Samples/BlazorUnitedApp/Pages/FetchData.razor
Comment (SteveSandersonMS) [inline src/Components/Samples/BlazorUnitedApp/Pages/FetchData.razor:9]: Is there some limitation that makes this necessary? What would happen if the table was inlined into `FetchData.razor`, and that was marked with `@attribute [RenderModeServer]`?

File: src/Components/Web.JS/src/Boot.Web.ts
Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Boot.Web.ts:43]: Splitting the artifacts into independently-loadable ES modules would be a great benefit. It's probably quite a big extra challenge as we have to think hard about how granularly to split things up so we can limit the amount of overlap (code loaded twice) when both Server and WebAssembly turn out to be used, given the efficiency/simplicity tradeoffs.

While it would be great to do this during .NET 8, I expect we can consider it a relatively low priority as we'll still have a fully-functional product without it.

One possible starting point that sidesteps most of the complexity would be something like:

 * `blazor.web.js`: only includes SSR support (streaming and progressively-enhanced nav, etc.)
   * (Some other child module that gets loaded on demand): includes all the interactive pieces for Server and WebAssembly

This would not involve solving many hard problems since most of those are due to the fine-grained overlap between Server and WebAssembly, and yet still delivers on the promise that SSR can be super lightweight.

Either way this PR is already pretty heavy on refactoring so it would probably be advantageous to save that for later. There is already a list of "todo" items associated with this PR description.

File: src/Components/WebAssembly/WebAssembly/src/Rendering/WebAssemblyRenderer.cs
Comment (SteveSandersonMS) [inline src/Components/WebAssembly/WebAssembly/src/Rendering/WebAssemblyRenderer.cs:57]: Very good to see this being removed.

File: src/Components/Web/src/WebRenderer.cs
Comment (SteveSandersonMS) [reply src/Components/Web/src/WebRenderer.cs:43]: Yeah this is a bit awkward. There's nothing objectively incorrect about starting a task from a constructor, but it definitely pushes up against the boundaries of normal constructor behavior.

The factory method option is certainly a viable possibility, albeit with its own drawbacks (if you're not supposed to use the constructor any more, this becomes a viral concern since the same would be true for any derived classes). Right now we could get away with the factory method as only a minor breaking change.

However @javiercn's points about other approaches are also promising. Do we really need to support registering an arbitrary number of renderers dynamically? What limitations would we face if we just said there would only be at most one of a given type, and we have some convention like 0=server, 1=webassembly? Is it enough just for these two type not to clash?

Comment (javiercn) [reply src/Components/Web/src/WebRenderer.cs:43]: Actually, you bring up a good point. We can use a HILO strategy, where each mode gets a different "HI" value, and we use different LO values. To be clear, if the render mode is an Int32, take the two most significant bytes and give 0 to Server and 1 to Webassembly, (2 to Webview), and use 0 for the two least significant bytes. If in the future we want to support multiple renderers of the same type, each one will just "rev" the ID.

File: src/Components/Samples/BlazorUnitedApp/Program.cs
Comment (MackinnonBuck) [reply src/Components/Samples/BlazorUnitedApp/Program.cs:32]: Definitely, my [plan](https://github.com/dotnet/aspnetcore/pull/48294#discussion_r1197202826) was to revert changes to the sample. The changes were just to make testing convenient.

File: src/Components/Web.JS/src/Boot.Server.Common.ts
Comment (MackinnonBuck) [reply src/Components/Web.JS/src/Boot.Server.Common.ts:76]: Just so I'm understanding correctly, is the only change you're suggesting that we change it from `DotNet.attachDispatcher(...)` to something like `new DotNet.CallDispatcher(...)`?

There already is a server and client implementation with the way things are currently set up. `attachDispatcher()` new-s up a new `CallDispatcher` instance, plus does some additional checks like "if this is the first dispatcher, make it the default" and "if this is not the first dispatcher, remove the default".

I guess the one advantage of keeping the `attachDispatcher()` way of doing things is that it can return an `ICallDispatcher`, which has a more restrictive public API compared to the non-exported `CallDispatcher` implementation.

File: src/Components/Web.JS/src/Boot.Web.ts
Comment (MackinnonBuck) [reply src/Components/Web.JS/src/Boot.Web.ts:43]: All that sounds good to me. Let's save this item for a follow-up PR.

File: src/Components/Samples/BlazorUnitedApp/Pages/FetchData.razor
Comment (MackinnonBuck) [reply src/Components/Samples/BlazorUnitedApp/Pages/FetchData.razor:9]: Nope, this isn't necessary - that scenario works fine. I was just modifying the sample project to validate various scenarios, and the changes will be reverted before this gets merged.

File: src/Components/Web.JS/src/Rendering/WebRendererInteropMethods.ts
Comment (MackinnonBuck) [reply src/Components/Web.JS/src/Rendering/WebRendererInteropMethods.ts:18]: Yeah, this will require some more thinking. I'm going to resolve this conversation for now - I think this can be addressed in a follow-up PR.

File: src/Components/Web/src/WebRenderer.cs
Comment (MackinnonBuck) [reply src/Components/Web/src/WebRenderer.cs:43]: > If we want to avoid this dance, we can change the Renderer ID for a GUID instead of an int.

This could work, but wouldn't it mean we're serializing and sending a much larger ID to JavaScript each time we render?

> What limitations would we face if we just said there would only be at most one of a given type, and we have some convention like 0=server, 1=webassembly? Is it enough just for these two type not to clash?

This was the approach I took during prototyping, and it works, but I decided not to do that in this PR because:
* We have to keep those hard-coded IDs in sync between .NET and JS code
* We have to remember to assign new a renderer ID to each execution model we add in the future
* It makes it more difficult to enable having multiple renderers per platform, in case that ever becomes useful (although I recognize Javier's HILO suggestion solves that)

> We can use a HILO strategy, where each mode gets a different "HI" value, and we use different LO values. To be clear, if the render mode is an Int32, take the two most significant bytes and give 0 to Server and 1 to Webassembly, (2 to Webview), and use 0 for the two least significant bytes. If in the future we want to support multiple renderers of the same type, each one will just "rev" the ID.

I'm not sure I'm totally convinced yet that these alternate approaches are better than what we currently have. Even if we were to change the way to allocate renderer IDs, we would still have to call `attachWebRendererInterop` from .NET to provide JS with the dependencies it needs to perform rendering. So, we'd still have the problem of figuring out the right way make that invocation; it would still be problematic to attempt to render before that call completes, even if the .NET side doesn't rely on the return value of `attachWebRendererInterop` to assign a renderer ID. If we have to do the JS interop call anyway, what are we losing by using it to also allocate a renderer ID?

I do totally agree that it's not ideal for the `WebRenderer` constructor to initiate a JS interop call. I think the factory approach would be the best way to handle it, but given that it would be a breaking change (and that we've been invoking the JS interop call in the constructor for a while now already), it seems fine to me to keep things as they are.

Thoughts?

Comment (SteveSandersonMS) [reply src/Components/Web/src/WebRenderer.cs:43]: > wouldn't it mean we're serializing and sending a much larger ID to JavaScript each time we render?

Yes, would be nice to avoid that.

> and that we've been invoking the JS interop call in the constructor for a while now already

That is a good point - I agree that since this change doesn't actually worsen anything, it's not really necessary to force an architectural or API-breaking change here. Currently this is only an internal detail and if we want to make an API-breaking change to turn this into a factory, there's nothing better about doing that now than any time in the future.

We can see if @javiercn has further practical reasons for changing this but otherwise this looks good to me.

File: src/Components/Samples/BlazorUnitedApp.WebAssembly/Program.cs
Comment (SteveSandersonMS) [reply src/Components/Samples/BlazorUnitedApp.WebAssembly/Program.cs]: It's removed now anyway so will mark as resolved.

**Review by SteveSandersonMS — APPROVED** [review]
Looks great to me! Would be best to get an ack from @javiercn about the renderer ID thing before merging if possible.

**Review by javiercn — APPROVED** [review]

**Pull request merged**
